src/dmd/errors.d

```diff
@@ -401,19 +401,49 @@ extern (C++) void message(const(char)* format, ...)
 }
 
 /**
- * Just print to stderr, doesn't care about gagging.
- * (format,ap) text within backticks gets syntax highlighted.
- * Params:
- *      loc         = location of error
- *      headerColor = color to set `header` output to
- *      header      = title of error message
- *      format      = printf-style format specification
- *      ap          = printf-style variadic arguments
- *      p1          = additional message prefix
- *      p2          = additional message prefix
- */
+* To convert a function pointer into a delegate.
+*/
+auto toDelegate(F)(auto ref F func)
+{
+    static if (is(typeof(*F) Params == __parameters))
+    {
+        return (Params params) {
+            return func(params);
+        };
+    }
+    else
+    {
+        static assert(false, "The given argument `func` is not a function type");
+    }
+}
+
+alias diagnosticHandlerDelegate = void delegate(const ref Loc, Color, const(char)*, const(char)*, va_list, const(char)*, const(char)*) nothrow;
+diagnosticHandlerDelegate diagnosticHandler;    //a delegate to handle what to do with diagnostics
+
+/**
+* It calls diagnosticHandler to handle diagnostics.
+*/
 private void verrorPrint(const ref Loc loc, Color headerColor, const(char)* header,
         const(char)* format, va_list ap, const(char)* p1 = null, const(char)* p2 = null)
+{
+    diagnosticHandler(loc,headerColor, header, format, ap, p1, p2);
```

Comment:
We can delegate the addition of toDelegate to the caller and remove non-default initialization if we just add
```
if(diagnosticHandler)
    return diagnosticHandler(loc, headerColor, header, format, ap, p1, p2);
```
and otherwise continue with the existing implementation of `verrorPrint`.